notebooks/WFC3/point_spread_function/psf_utilities.py

```diff
@@ -0,0 +1,402 @@
+                print('x_shift, y_shift =', x_shift, y_shift)
+
+            '''
+            Calculate the center of mass of the inner 5 pixels similar to hst1pass.
+            This provides a centroid for comparison with the user provided values.
+            In general, the values should be very close to rpix but not equal exactly.
+            Shift the image by the sub-pixel offset for sub-pixel alignment accuracy.
+            '''
+            
+            mask = numpy.array([[True, False, True], [False, False, False], [True, False, True]])
+            xcom, ycom = centroid_com(subimage[rpix-1:rpix+2, rpix-1:rpix+2], mask=mask)
+            if (verbose is True):
+                print('Before shift centroid =', round(xcom+rpix-1, 4), round(ycom+rpix-1, 4))
+            my_shift = [y_shift, x_shift]
+            subimage = shift(subimage, my_shift, mode='mirror')
+            xcom, ycom = centroid_com(subimage[rpix-1:rpix+2, rpix-1:rpix+2], mask=mask)
+            if (verbose is True):
+                print('After shift centroid = ', round(xcom+rpix-1, 4), round(ycom+rpix-1, 4))
+        else:
+            if (verbose is True):
+                print('Warning: The function is not aligning at the sub-pixel level. Is this intended?')
+
+        # # If desired we can recentroid. This would be false using hst1pass and true for approximate centroids.
+        # if (recentroid is True):
+        #     print('Recalculating centroid based on 5-pixel center of mass.')
+        #     print('Using:', subimage[rpix-1:rpix+2, rpix-1:rpix+2])
+        #     mask=numpy.array([[True, False, True], [False, False, False], [True, False, True]]) # true = masked.
+        #     print('Before shift centroid =', centroid_com(subimage[rpix-1:rpix+2, rpix-1:rpix+2], mask=mask))
+        #     #my_shift = [y_shift, x_shift] # ensure correct positive and negative.
+        #     #subimage = scipy.ndimage.shift(subimage, my_shift, mode='mirror')
+        #     print('After shift centroid = ', centroid_com(subimage[rpix-1:rpix+2, rpix-1:rpix+2], mask=mask)) # Should be very very close to rpix but not equal exactly.
+        
+        # Determine the location of the maximum flux.
+        peak_location = numpy.unravel_index(numpy.argmax(subimage, axis=None), subimage.shape)
+        if (verbose is True):
+            print('The subimage peak flux (x,y) = (' + str(peak_location[1]) + ', ' + str(peak_location[0]) + ')')
+        
+        # Scale to maximum flux so all stars peak at unity.
+        if (scale_stars is True):
+            print('Scaling the stars peak flux to unity...')
+            subimage = (subimage/numpy.amax(subimage))
+            
+        # Protect against peak_finder results that do not contain a star.
+        if (peak_location[1] == 0 and peak_location[0] == 0):
+            print('This object is outside the data region and will be excluded.\n')
+            
+        if (peak_location[1] != 0 and peak_location[0] != 0):
```

Comment:
Are there other sub conditions (e.g. `if peak_location[1] == 0 and peak_location[0] != 0`)? If not, using `else` should work here

Reply:
Unfortunately this is a bit complex. If all is well and the bright star is centered then the peak will be at (rpix, rpix). However, if the peak finder fails, it returns (0,0). However, it's also possible to have another bright star in the field so that the peak is not at (rpix, rpix), but the star can still be appended if a median is being taken. In this case I've combined them into an if-else statement, and users can modify if they'd like to reject sources with other bright stars.